### PR TITLE
Automated dependency tree update

### LINKED ISSUE #837: When upgrading dependency hash, fpm does not rebuild
When I upgrade the git hash of a dependency in `fpm.toml`, my current project does not rebuild, and consequently the new dependency is not used. Fpm must read the hash of the dependency into the model, and if it changes, it must rebuild.

### PR DESCRIPTION
**Pull request by perazz**
Fixes #837 #799 (also targeting #838).

tl;dr fpm does never update a project's cached dependencies unless forced to do so via `fpm update`. 
This is wrong as dependencies may change in the manifest and the user should not have to manually do it every time. 

With the current PR: 

1) The ability to update dependencies is propagated to `fpm build`:
https://github.com/perazz/fpm/blob/063148259fc4745f60c8e3d253c803b3e09ace38/src/fpm.f90#L63-L70 

2) When `cache.toml` is present, the old dependency tree was only loaded from it. New dependencies would be added from scratch (id==0), but cached dependencies (id>0) would be left as-is. Now, instead of just finding the ID of a cached dependency, fpm will check if it has changed, and request an update if so
https://github.com/perazz/fpm/blob/063148259fc4745f60c8e3d253c803b3e09ace38/src/fpm/dependency.f90#L373-L389

This resolves changes in `fpm.toml` automatically and should be easily integrated with the local registry cache (#817 @minhqdao). @certik please let me know about the module contents problems (#838), I'm happy to help.

### CODE REVIEW

**Review by henilp105 — APPROVED** [review]
Thanks @perazz , look good to me.
